certbot/tests/crypto_util_test.py

```diff
@@ -237,6 +240,101 @@ def test_it(self):  # pylint: disable=no-self-use
+        self.renewable_cert.privkey = RSA512_KEY_PATH
+        self.renewable_cert.fullchain = test_util.vector_path('self_signed_fullchain.pem')
+
+        self.bad_renewable_cert = mock.MagicMock()
+        self.bad_renewable_cert.chain = SS_CERT_PATH
+        self.bad_renewable_cert.cert = SS_CERT_PATH
+        self.bad_renewable_cert.fullchain = SS_CERT_PATH
+
+
+class VerifyRenewableCertTest(VerifyCertSetup):
+    """Tests for certbot.crypto_util.verify_renewable_cert."""
+
+    def setUp(self):
+        super(VerifyRenewableCertTest, self).setUp()
+
+    def _call(self, renewable_cert):
+        from certbot.crypto_util import verify_renewable_cert
+        return verify_renewable_cert(renewable_cert)
+
+    def test_verify_renewable_cert(self):
+        self.assertEqual(None, self._call(self.renewable_cert))
+
+    @mock.patch('certbot.crypto_util.verify_renewable_cert_sig', side_effect=errors.Error(""))
+    def test_verify_renewable_cert_failure(self, unused_verify_renewable_cert_sign):
+        self.assertRaises(errors.Error, self._call, self.bad_renewable_cert)
+
+
+class VerifyRenewableCertSigTest(VerifyCertSetup):
+    """Tests for certbot.crypto_util.verify_renewable_cert."""
+
+    def setUp(self):
+        super(VerifyRenewableCertSigTest, self).setUp()
+
+    def _call(self, renewable_cert):
+        from certbot.crypto_util import verify_renewable_cert_sig
+        return verify_renewable_cert_sig(renewable_cert)
+
+    def test_cert_sig_match(self):
+        self.assertEqual(None, self._call(self.renewable_cert))
+
+    def test_cert_sig_mismatch(self):
+        self.bad_renewable_cert.cert = test_util.vector_path('self_signed_cert_bad.pem')
+        self.assertRaises(errors.Error, self._call, self.bad_renewable_cert)
+
+
+class VerifyFullchainTest(VerifyCertSetup):
+    """Tests for certbot.crypto_util.verify_fullchain."""
+
+    def setUp(self):
+        super(VerifyFullchainTest, self).setUp()
+
+    def _call(self, renewable_cert):
+        from certbot.crypto_util import verify_fullchain
+        return verify_fullchain(renewable_cert)
+
+    def test_fullchain_matches(self):
+        self.assertEqual(None, self._call(self.renewable_cert))
+
+    def test_fullchain_mismatch(self):
+        self.assertRaises(errors.Error, self._call, self.bad_renewable_cert)
+
+    def test_fullchain_ioerror(self):
+        self.bad_renewable_cert.chain = "dog"
+        self.assertRaises(errors.Error, self._call, self.bad_renewable_cert)
+
+
+class VerifyCertMatchesPrivKeyTest(VerifyCertSetup):
+    """Tests for certbot.crypto_util.verify_cert_matches_priv_key."""
+
+    def setUp(self):
+        super(VerifyCertMatchesPrivKeyTest, self).setUp()
+
+    def _call(self, renewable_cert):
+        from certbot.crypto_util import verify_cert_matches_priv_key
+        return verify_cert_matches_priv_key(renewable_cert)
+
+    def test_cert_priv_key_match(self):
+        self.assertEqual(None, self._call(self.renewable_cert))
```

Comment:
Ugh. Want to know something fun? If you run these tests with a newer version of OpenSSL (I used OpenSSL 1.1.0e in Debian testing), this test (as well as `VerifyRenewableCertTest.test_verify_renewable_cert`) fail with:
```
Error: verifying the cert located at /root/certbot/certbot/tests/testdata/self_signed_cert.pem matches the                 private key located at /root/certbot/certbot/tests/testdata/rsa512_key.pem has failed.                 Details: [('SSL routines', 'SSL_CTX_use_certificate', 'ee key too small')]
```
This will be a problem for OS package maintainers using newer OpenSSL versions. The production code or tests don't have to change, but we should probably generate a newer self-signed cert that uses a larger RSA key. I tested with a 2048 bit key and tests pass. I didn't see if we could get away with a smaller 1024 bit key.

If you'd like help testing this or generating these certs/keys, let me know!